File: examples/sorting-algorithms/sut-merge-sort.ts
/**
 * Merge Sort SUT
 *
 * Recursive divide-and-merge sort. Each comparison during merge is counted.
 * A swap is counted each time a right-side element is placed before a
 * left-side element during the merge step.
 */

interface SortInput {
	data: number[];
}

interface SortResult {
	sorted: number[];
	comparisons: number;
	swaps: number;
	executionTimeMs: number;
}

interface Counters {
	comparisons: number;
	swaps: number;
}

function merge(
	left: number[],
	right: number[],
	counters: Counters,
): number[] {
	const result: number[] = [];
	let i = 0;
	let j = 0;

	while (i < left.length && j < right.length) {
		counters.comparisons++;
		if (left[i] <= right[j]) {
			result.push(left[i]);
			i++;
		} else {
			result.push(right[j]);
			counters.swaps++;
			j++;
		}
	}

	while (i < left.length) {
		result.push(left[i]);
		i++;
	}

	while (j < right.length) {
		result.push(right[j]);
		j++;
	}

	return result;
}

function mergeSort(arr: number[], counters: Counters): number[] {
	if (arr.length <= 1) {
		return arr;
	}

	const mid = Math.floor(arr.length / 2);
	const left = mergeSort(arr.slice(0, mid), counters);
	const right = mergeSort(arr.slice(mid), counters);

	return merge(left, right, counters);
}

export function createSut(): {
	id: string;
	config: Record<string, unknown>;
	run: (input: SortInput) => Promise<SortResult>;
} {
	return {
		id: "merge-sort",
		config: {},
		run: async (input: SortInput): Promise<SortResult> => {
			const arr = [...input.data];
			const counters: Counters = { comparisons: 0, swaps: 0 };

			const start = performance.now();
			const sorted = mergeSort(arr, counters);
			const executionTimeMs = performance.now() - start;

			return {
				sorted,
				comparisons: counters.comparisons,
				swaps: counters.swaps,
				executionTimeMs,
			};
		},
	};
}

File: examples/sorting-algorithms/sut_merge_sort.py
"""Merge Sort SUT — recursive divide-and-merge."""

import time


def create_sut():
    def run(input):
        arr = list(input["data"])
        counters = {"comparisons": 0, "swaps": 0}

        def merge(left, right):
            result = []
            i = j = 0
            while i < len(left) and j < len(right):
                counters["comparisons"] += 1
                if left[i] <= right[j]:
                    result.append(left[i])
                    i += 1
                else:
                    result.append(right[j])
                    counters["swaps"] += 1
                    j += 1
            result.extend(left[i:])
            result.extend(right[j:])
            return result

        def merge_sort(a):
            if len(a) <= 1:
                return a
            mid = len(a) // 2
            left = merge_sort(a[:mid])
            right = merge_sort(a[mid:])
            return merge(left, right)

        start = time.perf_counter()
        sorted_arr = merge_sort(arr)
        execution_time_ms = (time.perf_counter() - start) * 1000

        return {
            "sorted": sorted_arr,
            "comparisons": counters["comparisons"],
            "swaps": counters["swaps"],
            "executionTimeMs": execution_time_ms,
        }

    return {"id": "merge-sort", "config": {}, "run": run}
